interrupt properly each run through channel (and side effect fix compilation error and works on windows)

ldemailly opened the pull request:
fixes #63

ldemailly committed:
fix compilation error

but still not working
partial fix for #63

ldemailly committed:
Updated comment and fixed lint

ldemailly committed:
Now use separate channel to interrupt runs

cleaner and bonus, works on windows

[wip will add per run interrupt]

ymesika commented:
@ldemailly I'm using Mac and can't verify that it indeed fixed the compilation error on Windows

ldemailly committed:
Now can interrupt specific/correct run for each window/UI

ldemailly commented:
@ymesika welcome back ! it's not as much about windows than about per run channel now, PTAL

ldemailly committed:
Use the variable

ldemailly committed:
Undoing 1 too many exit->stop rename

ldemailly committed:
Allow save+json to both work together

ymesika commented on ui/uihandler.go:
```diff
@@ -101,10 +104,11 @@ func (w *HTMLEscapeWriter) Write(p []byte) (int, error) {
-	if r.FormValue("exit") == "Exit" {
-		log.Critf("Exit request from %v", r.RemoteAddr)
-		DoExit = true
+	DoStop := false
```
Seems like you can't actually get a form with both 'stop' and 'load' set together? In this case maybe DoStop and DoLoad can be combined just to hold request type.

ldemailly replied:
changed to enum, thanks!

ymesika commented on ui/uihandler.go:
```diff
@@ -231,6 +264,9 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		if err != nil {
 			log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
 		}
+		if DoSave {
```
Adding one more type to the request type variable I mentioned in my previous comment - stop / load / save

ldemailly replied:
well not save as save is that checkbox is in addition to load / json

ymesika reviewed (approved):
Beside my comments below, which are cosmetics, I tested this new button and it works fine.
Wasn't able to check if it indeed got rid of the compilation error on Windows.

ldemailly commented on periodic/periodic.go:
```diff
@@ -59,14 +59,14 @@ type Runnable interface {
-	if len(ro.Runners) < ro.NumThreads {
-		log.Infof("Resizing runners from %d to %d", len(ro.Runners), ro.NumThreads)
-		ro.Runners = make([]Runnable, ro.NumThreads)
+func (r *RunnerOptions) MakeRunners(rr Runnable) {
```
this change is to minimize the diff for Normalize (as go linters want the variable name for receivers to be all the same)

ldemailly committed:
Changed the signal handling to be fully correct

Also code review comment in UI handler, using a mode enum and a switch
(it’s only marginally more readable)

ldemailly committed:
It doesn't leak anymore

ldemailly committed:
UI should also use lock to avoid panic on already closed channel

Also need to use a copy of the channel in loops to not wait on nil

rkpagadala commented on ui/uihandler.go:
```diff
@@ -155,98 +194,106 @@ func Handler(w http.ResponseWriter, r *http.Request) {
+		} else { // Stop one
+			uiRunMapMutex.Lock()
+			v := runs[runid]
+			if v != nil {
```
Change to
v, ok := runs[runid]
 			if ok {

ldemailly replied:
k

rkpagadala commented on periodic/periodic.go:
```diff
@@ -128,9 +132,15 @@ type periodicRunner struct {
-func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
-	r := &periodicRunner{*opts} // by default just copy the input params
+var (
+	globalAbort     chan os.Signal
```
Consider changing to
gAbortChan

ldemailly replied:
k

rkpagadala commented on periodic/periodic.go:
```diff
@@ -128,9 +132,15 @@ type periodicRunner struct {
-	r := &periodicRunner{*opts} // by default just copy the input params
+var (
+	globalAbort     chan os.Signal
+	outstandingRuns int64
```
gOutstandingRuns

ldemailly replied:
k

rkpagadala commented on periodic/periodic.go:
```diff
@@ -128,9 +132,15 @@ type periodicRunner struct {
+var (
+	globalAbort     chan os.Signal
+	outstandingRuns int64
+	abortMutex      sync.Mutex
```
gAbortMutex

Or put then all in a structure and call it gSomething

ldemailly replied:
k

rkpagadala reviewed (approved):
Minor comments

ldemailly committed:
Add windows compile check + code review comments + lint

rkpagadala commented:
/lgtm
/approve
